src/framework/components/joint/component.js

```diff
@@ -0,0 +1,371 @@
+            this._angularLimitsY.copy(limits);
+            this._updateAngularLimits();
+        }
+    }
+
+    get angularLimitsY() {
+        return this._angularLimitsY;
+    }
+
+    set angularLimitsZ(limits) {
+        if (!this._angularLimitsZ.equals(limits)) {
+            this._angularLimitsZ.copy(limits);
+            this._updateAngularLimits();
+        }
+    }
+
+    get angularLimitsZ() {
+        return this._angularLimitsZ;
+    }
+
+    set linearLimitsX(limits) {
+        if (!this._linearLimitsX.equals(limits)) {
+            this._linearLimitsX.copy(limits);
+            this._updateLinearLimits();
+        }
+    }
+
+    get linearLimitsX() {
+        return this._linearLimitsX;
+    }
+
+    set linearLimitsY(limits) {
+        if (!this._linearLimitsY.equals(limits)) {
+            this._linearLimitsY.copy(limits);
+            this._updateLinearLimits();
+        }
+    }
+
+    get linearLimitsY() {
+        return this._linearLimitsY;
+    }
+
+    set linearLimitsZ(limits) {
+        if (!this._linearLimitsZ.equals(limits)) {
+            this._linearLimitsZ.copy(limits);
+            this._updateLinearLimits();
+        }
+    }
+
+    get linearLimitsZ() {
+        return this._linearLimitsZ;
+    }
+
+    _convertTransform(pcTransform, ammoTransform) {
+        const pos = pcTransform.getTranslation();
+        const rot = new Quat();
+        rot.setFromMat4(pcTransform);
+
+        const ammoVec = new Ammo.btVector3(pos.x, pos.y, pos.z);
+        const ammoQuat = new Ammo.btQuaternion(rot.x, rot.y, rot.z, rot.w);
+
+        ammoTransform.setOrigin(ammoVec);
+        ammoTransform.setRotation(ammoQuat);
+
+        Ammo.destroy(ammoVec);
+        Ammo.destroy(ammoQuat);
+    }
+
+    _updateAngularLimits() {
+        const constraint = this._constraint;
+        if (constraint) {
+            const alx = this._angularLimitsX;
+            const aly = this._angularLimitsY;
+            const alz = this._angularLimitsZ;
+
+            const limits = new Ammo.btVector3(alx.x, aly.x, alz.x);
+            constraint.setAngularLowerLimit(limits);
+            limits.setValue(alx.y, aly.y, alz.y);
+            constraint.setAngularUpperLimit(limits);
+            Ammo.destroy(limits);
+        }
+    }
+
+    _updateLinearLimits() {
+        const constraint = this._constraint;
+        if (constraint) {
+            const llx = this._linearLimitsX;
+            const lly = this._linearLimitsY;
+            const llz = this._linearLimitsZ;
+
+            const limits = new Ammo.btVector3(llx.x, lly.x, llz.x);
+            constraint.setLinearLowerLimit(limits);
+            limits.setValue(llx.y, lly.y, llz.y);
+            constraint.setLinearUpperLimit(limits);
+            Ammo.destroy(limits);
+        }
+    }
+
+    _createConstraint() {
+        if (this._entityA && this._entityA.rigidbody) {
+            this._destroyConstraint();
+
+            const mat = new Mat4();
```

Comment:
avoid allocation?

Reply:
Again, it's a case of slow down 99.9% app's engine init time or incur a tiny cost for 0.1% of apps using joints. Let's leave it for now and adjust later if necessary.